File: libraries/uava/src/main/java/br/com/vitorsalgado/androidstarter/uava/rx/RxBus.java
package br.com.vitorsalgado.androidstarter.uava.rx;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.subjects.PublishSubject;

public final class RxBus {
	private static final RxBus INSTANCE = new RxBus();
	private final PublishSubject<Object> mEventBus = PublishSubject.create();

	public static RxBus getInstance() {
		return INSTANCE;
	}

	public void send(final Object event) {
		mEventBus.onNext(event);
	}

	@SuppressWarnings("unchecked")
	public <T> Disposable register(final Class<T> eventClass, DisposableObserver<T> onNext) {
		return mEventBus
				.filter(o -> eventClass.isAssignableFrom(o.getClass()))
				.map(o -> (T) o)
				.subscribeWith(onNext);
	}

	public Observable<Object> toObservable() {
		return mEventBus;
	}

	public boolean hasObservers() {
		return mEventBus.hasObservers();
	}
}

File: libs/utils/src/main/kotlin/br/com/vitorsalgado/example/utils/RxBus.kt
package br.com.vitorsalgado.example.utils

import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject

object RxBus {
  val disposables = mutableMapOf<Any, List<CompositeDisposable>>()
  val publisher: PublishSubject<Any> = PublishSubject.create()


  fun publish(event: Any) = publisher.onNext(event)

  fun <T> subscribe(eventType: Class<T>): Observable<T> = publisher.ofType(eventType)

  fun <T> register(eventType: Class<T>, subscription: CompositeDisposable) {
    val list = getDisposablesByEvent(eventType).apply {
      add(subscription)
    }

    disposables[eventType] = list
  }

  private fun <T> getDisposablesByEvent(eventType: Class<T>): MutableList<CompositeDisposable> = disposables[eventType].orEmpty().toMutableList()
}
